### PR TITLE
replace upstream Dockerfiles with ubi dockerfiles.

### PR DESCRIPTION
**Pull request by VedantMahabaleshwarkar**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://issues.redhat.com/browse/RHOAIENG-6506  

`Note: Replaced existing Dockerfiles because that file is what Konflux uses to build the images. Also, this will help track any changes to the upstream dockerfile because it will create a merge-conflict (as opposed to maintaining a separate `Dockerfile.ubi8` file). 

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- tested `podman build` succeeds locally.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Build the manager binary
-FROM golang:1.21 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.19 as builder
```
Comment (spolti) [inline Dockerfile:2]: kserve is based on go1.21, I don't think a good idea move it back.
it will probably fail to build.

File: Dockerfile [hunk 2]
```diff
@@ -14,8 +14,8 @@ COPY pkg/    pkg/
 # Build
 RUN CGO_ENABLED=0 GOOS=linux GOFLAGS=-mod=mod go build -a -o manager ./cmd/manager
 
-# Copy the controller-manager into a thin image
-FROM gcr.io/distroless/static:nonroot
+# Use distroless as minimal base image to package the manager binary
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.6
```
Comment (spolti) [inline Dockerfile:18]: latest?

File: Dockerfile (hunk 1 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile]: updated

File: Dockerfile (hunk 2 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile:18]: is it a good idea to use floating tags? I thought the best practice is to avoid using `latest` is it not?

Comment (Jooho) [reply Dockerfile:18]: it should be `latest`

File: Dockerfile [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # Build the manager binary
-FROM golang:1.21 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.21 as builder
```
Comment (Jooho) [inline Dockerfile:2]: ubi9 is for 1.21 but filippe noticed some python versions with it. @spolti is it ok to go with ubi9 1.21?

File: router.Dockerfile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # Build the inference-router binary
-FROM golang:1.21 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.21 as builder
```
Comment (Jooho) [inline router.Dockerfile:2]: please see the above comment.

File: agent.Dockerfile [hunk 5]
```diff
@@ -8,15 +8,15 @@ COPY go.sum  go.sum
...
-COPY cmd/    cmd/
 COPY pkg/    pkg/
+COPY cmd/    cmd/
 
 # Build
 RUN CGO_ENABLED=0 GOOS=linux GOFLAGS=-mod=mod go build -a -o agent ./cmd/agent
 
 # Copy the inference-agent into a thin image
-FROM gcr.io/distroless/static:nonroot
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.6
```
Comment (Jooho) [inline agent.Dockerfile:18]: please see the above comment.

File: router.Dockerfile [hunk 6]
```diff
@@ -8,15 +8,15 @@ COPY go.sum  go.sum
...
-COPY cmd/    cmd/
 COPY pkg/    pkg/
+COPY cmd/    cmd/
 
 # Build
 RUN CGO_ENABLED=0  go build -a -o router ./cmd/router
 
 # Copy the inference-router into a thin image
-FROM gcr.io/distroless/static:nonroot
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.6
```
Comment (Jooho) [inline router.Dockerfile:18]: please see the above comment.

**Comment by israel-hdez** [comment]
We also should update the storage initializer one.

File: Dockerfile (hunk 3 above)
Comment (spolti) [reply Dockerfile:2]: idk, it needs to be tested.
I only tested rest-proxy which required some tweaks. Not sure how things will go here..

**Comment by spolti** [comment]
> We also should update the storage initializer one.

If I am not mistake, custo_transformers and custom_model as well.

File: Dockerfile (hunk 2 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile]: fixed

File: Dockerfile (hunk 3 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile:2]: the latest version of the code builds successfully locally.

**Comment by VedantMahabaleshwarkar** [comment]
> We also should update the storage initializer one.  


@israel-hdez added the storage initializer dockerfile as well. 

>We also should update the storage initializer one.
If I am not mistake, custo_transformers and custom_model as well.


@spolti doesn't look like we use/build those today https://github.com/opendatahub-io/kserve/blob/master/config/overlays/odh/params.env

File: Dockerfile (hunk 3 above)
Comment (spolti) [reply Dockerfile:2]: nice, they have updated the ubi8 go-toolset

File: python/storage-initializer.Dockerfile [hunk 7]
```diff
@@ -1,45 +1,33 @@
...
 
-FROM ${BASE_IMAGE} as builder
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6 as builder
+
+# Install Python and dependencies
+RUN microdnf install -y python39 python39-devel gcc libffi-devel openssl-devel && microdnf clean all
 
 # Install Poetry
 ARG POETRY_HOME=/opt/poetry
 ARG POETRY_VERSION=1.7.1
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:11]: use the latest version, 1.8.3

File: python/storage-initializer.Dockerfile [hunk 8]
```diff
@@ -1,45 +1,33 @@
...
-    krb5-config \
-    && rm -rf /var/lib/apt/lists/*
-
 RUN pip install --no-cache-dir krbcontext==0.10 hdfs~=2.6.0 requests-kerberos==0.14.0
-# Fixes Quay alert GHSA-2jv5-9r88-3w3p https://github.com/Kludex/python-multipart/security/advisories/GHSA-2jv5-9r88-3w3p
 RUN pip install --no-cache-dir starlette==0.36.2
 
-
-FROM ${BASE_IMAGE} as prod
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6 as prod
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:30]: latest.

File: Dockerfile [hunk 9]
```diff
@@ -12,10 +12,13 @@ COPY cmd/    cmd/
...
 RUN CGO_ENABLED=0 GOOS=linux GOFLAGS=-mod=mod go build -a -o manager ./cmd/manager
 
-# Copy the controller-manager into a thin image
-FROM gcr.io/distroless/static:nonroot
+# Use distroless as minimal base image to package the manager binary
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6
 COPY third_party/ /third_party/
 COPY --from=builder /go/src/github.com/kserve/kserve/manager /
-ENTRYPOINT ["/manager"]
+USER 65532:65532
```
Comment (spolti) [inline Dockerfile:22]: any reason to use such a high number?

it is better to use the same approach than here:
```
RUN microdnf install -y shadow-utils && \
    microdnf clean all && \
    useradd kserve -m -u 1000 -d /home/kserve
```

File: Dockerfile (hunk 3 above)
Comment (Jooho) [reply Dockerfile:2]: nice timing :)

File: python/storage-initializer.Dockerfile [hunk 10]
```diff
@@ -1,45 +1,33 @@
...
-ARG DEBIAN_FRONTEND=noninteractive
-
-RUN apt-get update && apt-get install -y \
-    gcc \
-    libkrb5-dev \
-    krb5-config \
-    && rm -rf /var/lib/apt/lists/*
-
 RUN pip install --no-cache-dir krbcontext==0.10 hdfs~=2.6.0 requests-kerberos==0.14.0
-# Fixes Quay alert GHSA-2jv5-9r88-3w3p https://github.com/Kludex/python-multipart/security/advisories/GHSA-2jv5-9r88-3w3p
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:38]: do not remove it, to keep track why it is being installed manually here and tp remember to remove later when the kserve/project.toml brings the same or higher version.

File: python/storage-initializer.Dockerfile [hunk 11]
```diff
@@ -1,45 +1,33 @@
 ARG PYTHON_VERSION=3.9
-ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
 ARG VENV_PATH=/prod_venv
 
-FROM ${BASE_IMAGE} as builder
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6 as builder
```
Comment (Jooho) [inline python/storage-initializer.Dockerfile:4]: please use `latest` tag

File: python/storage-initializer.Dockerfile [hunk 12]
```diff
@@ -48,7 +36,9 @@ ARG VENV_PATH
 ENV VIRTUAL_ENV=${VENV_PATH}
 ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
-RUN useradd kserve -m -u 1000 -d /home/kserve
+RUN microdnf install -y shadow-utils && \
+    microdnf clean all && \
+    useradd kserve -m -u 1000 -d /home/kserve
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:41]: separate into the microdnf and user creation into two separate RUN instructions

File: router.Dockerfile [hunk 13]
```diff
@@ -8,15 +8,18 @@ COPY go.sum  go.sum
...
 RUN CGO_ENABLED=0  go build -a -o router ./cmd/router
 
 # Copy the inference-router into a thin image
-FROM gcr.io/distroless/static:nonroot
+FROM registry.access.redhat.com/ubi8/ubi-micro:latest
 COPY third_party/ third_party/
 WORKDIR /ko-app
 COPY --from=builder /go/src/github.com/kserve/kserve/router /ko-app/
-ENTRYPOINT ["/ko-app/router"]
+USER 65530:65530
```
Comment (spolti) [inline router.Dockerfile:23]: same here

File: python/storage-initializer.Dockerfile [hunk 14]
```diff
@@ -1,45 +1,33 @@
...
 
 COPY kserve/pyproject.toml kserve/poetry.lock kserve/
 RUN cd kserve && poetry install --no-root --no-interaction --no-cache --extras "storage"
 COPY kserve kserve
 RUN cd kserve && poetry install --no-interaction --no-cache --extras "storage"
 
-ARG DEBIAN_FRONTEND=noninteractive
-
-RUN apt-get update && apt-get install -y \
-    gcc \
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:32]: is these three packages available in the go-toolset image?

File: agent.Dockerfile [hunk 15]
```diff
@@ -8,15 +8,18 @@ COPY go.sum  go.sum
 
 RUN go mod download
 
-COPY cmd/    cmd/
 COPY pkg/    pkg/
+COPY cmd/    cmd/
 
 # Build
+USER root
```
Comment (Jooho) [inline agent.Dockerfile:15]: I wonder we need root user?

File: python/storage-initializer.Dockerfile [hunk 16]
```diff
@@ -1,45 +1,33 @@
...
-
-RUN python3 -m venv ${POETRY_HOME} && ${POETRY_HOME}/bin/pip install poetry==${POETRY_VERSION}
+RUN python -m venv ${POETRY_HOME} && ${POETRY_HOME}/bin/pip install poetry==${POETRY_VERSION}
 ENV PATH="$PATH:${POETRY_HOME}/bin"
 
 # Activate virtual env
 ARG VENV_PATH
 ENV VIRTUAL_ENV=${VENV_PATH}
-RUN python3 -m venv $VIRTUAL_ENV
+RUN python -m venv $VIRTUAL_ENV
```
Comment (spolti) [inline python/storage-initializer.Dockerfile:19]: make sure python points correctly to the python 3.9 binary.

@israel-hdez isn't python 3.10 the start version? 
need to check, honestly don't remember.

File: python/storage-initializer.Dockerfile (hunk 11 above)
Comment (spolti) [inline python/storage-initializer.Dockerfile:4]: latest

File: router.Dockerfile [hunk 17]
```diff
@@ -8,15 +8,18 @@ COPY go.sum  go.sum
 
 RUN go mod download
 
-COPY cmd/    cmd/
 COPY pkg/    pkg/
+COPY cmd/    cmd/
 
 # Build
+USER root
```
Comment (Jooho) [inline router.Dockerfile:15]: same question. To build go, do we need root user?

File: python/storage-initializer.Dockerfile (hunk 7 above)
Comment (VedantMahabaleshwarkar) [reply python/storage-initializer.Dockerfile:11]: latest version of poetry?

File: Dockerfile (hunk 9 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile:22]: followed the same pattern as the dockerfiles we have for odh-model-controller/modelmesh repos

File: python/storage-initializer.Dockerfile (hunk 16 above)
Comment (VedantMahabaleshwarkar) [reply python/storage-initializer.Dockerfile:19]: yep it does when I checked

File: agent.Dockerfile (hunk 15 above)
Comment (VedantMahabaleshwarkar) [reply agent.Dockerfile:15]: without root the build fails with permission error. We follow this pattern in modelmesh/odh-model-controller as well

File: router.Dockerfile (hunk 17 above)
Comment (VedantMahabaleshwarkar) [reply router.Dockerfile:15]: without root the build fails with permission error. We follow this pattern in modelmesh/odh-model-controller as well

File: python/storage-initializer.Dockerfile (hunk 14 above)
Comment (VedantMahabaleshwarkar) [reply python/storage-initializer.Dockerfile:32]: wouldn't the build have failed if the packages were missing?

File: python/storage-initializer.Dockerfile (hunk 10 above)
Comment (VedantMahabaleshwarkar) [reply python/storage-initializer.Dockerfile:38]: added the comment back

File: python/storage-initializer.Dockerfile (hunk 14 above)
Comment (spolti) [reply python/storage-initializer.Dockerfile:32]: gcc, yes, but kerberos, no.

Comment (VedantMahabaleshwarkar) [reply python/storage-initializer.Dockerfile:32]: added the kerberos packages. gcc is also included

File: Dockerfile (hunk 9 above)
Comment (VedantMahabaleshwarkar) [reply Dockerfile:22]: the `useradd` pattern led to some errors. The pattern in this PR (USER root -> build -> set different user at the end) is what other Dockerfiles for our components are following. @Jooho @spolti let me know if this is fine, IMO it's ok.

**Comment by VedantMahabaleshwarkar** [comment]
@spolti @Jooho all comments are addressed. PTAL

File: python/storage-initializer.Dockerfile [hunk 18]
```diff
@@ -1,45 +1,34 @@
...
-    krb5-config \
-    && rm -rf /var/lib/apt/lists/*
-
 RUN pip install --no-cache-dir krbcontext==0.10 hdfs~=2.6.0 requests-kerberos==0.14.0
 # Fixes Quay alert GHSA-2jv5-9r88-3w3p https://github.com/Kludex/python-multipart/security/advisories/GHSA-2jv5-9r88-3w3p
 RUN pip install --no-cache-dir starlette==0.36.2
 
-
-FROM ${BASE_IMAGE} as prod
+FROM registry.access.redhat.com/ubi8/ubi-minimal:latest as prod
```
Comment (israel-hdez) [inline python/storage-initializer.Dockerfile:31]: I don't think the storage-initializer will work. The ubi-minimal won't have python installed, so it won't run.

**Comment by VedantMahabaleshwarkar** [comment]
@Jooho I checked the CI failures, the failure is not in the test. The `destroy-cluster` sub step timed out after the test was successful. The ci failure can be ignored. PTAL.

**Comment by Jooho** [comment]
ci-fast show test passed but full ci failed
```
INFO[2024-05-30T18:59:12Z] Running step e2e-fast-e2e-fast.              
WARN[2024-05-30T19:16:52Z] Failed to get namespace ci-op-zvy0gbrz for heartbeating  error=Get "[https://172.30.0.1:443/api/v1/namespaces/ci-op-zvy0gbrz](https://172.30.0.1/api/v1/namespaces/ci-op-zvy0gbrz)": http2: client connection lost
INFO[2024-05-30T19:18:34Z] Step e2e-fast-e2e-fast succeeded after 19m21s. 
INFO[2024-05-30T19:18:34Z] Step phase test succeeded after 19m21s.
```
ci-slow show test passed but full ci failed
```
INFO[2024-05-30T19:09:18Z] Step e2e-slow-e2e-slow succeeded after 10m4s. 
INFO[2024-05-30T19:09:18Z] Step phase test succeeded after 10m4s.     
```
So I think we should ignore the ci failure.

**Pull request merged**